File: java_src/src/test/java/com/google/crypto/tink/signature/Ed25519ParametersTest.java
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.signature;

import static com.google.common.truth.Truth.assertThat;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public final class Ed25519ParametersTest {
  private static final Ed25519Parameters.Variant NO_PREFIX = Ed25519Parameters.Variant.NO_PREFIX;
  private static final Ed25519Parameters.Variant TINK = Ed25519Parameters.Variant.TINK;
  private static final Ed25519Parameters.Variant CRUNCHY = Ed25519Parameters.Variant.CRUNCHY;
  private static final Ed25519Parameters.Variant LEGACY = Ed25519Parameters.Variant.LEGACY;

  @Test
  public void buildParametersAndGetProperties_noPrefix() throws Exception {
    Ed25519Parameters parameters = Ed25519Parameters.create();
    assertThat(parameters.getVariant()).isEqualTo(NO_PREFIX);
    assertThat(parameters.hasIdRequirement()).isFalse();
  }

  @Test
  public void buildParametersAndGetProperties_noPrefix_setVariantExplicitly() throws Exception {
    Ed25519Parameters parameters = Ed25519Parameters.create(NO_PREFIX);
    assertThat(parameters.getVariant()).isEqualTo(NO_PREFIX);
    assertThat(parameters.hasIdRequirement()).isFalse();
  }

  @Test
  public void buildParametersAndGetProperties_tink() throws Exception {
    Ed25519Parameters parameters = Ed25519Parameters.create(TINK);
    assertThat(parameters.getVariant()).isEqualTo(TINK);
    assertThat(parameters.hasIdRequirement()).isTrue();
  }

  @Test
  public void buildParameterAndGetProperties_crunchy() throws Exception {
    Ed25519Parameters parameters = Ed25519Parameters.create(CRUNCHY);
    assertThat(parameters.getVariant()).isEqualTo(CRUNCHY);
    assertThat(parameters.hasIdRequirement()).isTrue();
  }

  @Test
  public void buildParameterAndGetProperties_legacy() throws Exception {
    Ed25519Parameters parameters = Ed25519Parameters.create(LEGACY);
    assertThat(parameters.getVariant()).isEqualTo(LEGACY);
    assertThat(parameters.hasIdRequirement()).isTrue();
  }

  @Test
  public void testEqualsAndEqualHashCode_noPrefix() throws Exception {
    Ed25519Parameters parametersNoPrefix0 = Ed25519Parameters.create();
    Ed25519Parameters parametersNoPrefix1 = Ed25519Parameters.create();
    assertThat(parametersNoPrefix0).isEqualTo(parametersNoPrefix1);
    assertThat(parametersNoPrefix0.hashCode()).isEqualTo(parametersNoPrefix1.hashCode());
  }

  @Test
  public void testEqualsAndEqualHashCode_tink() throws Exception {
    Ed25519Parameters parametersTink0 = Ed25519Parameters.create(TINK);
    Ed25519Parameters parametersTink1 = Ed25519Parameters.create(TINK);
    assertThat(parametersTink0).isEqualTo(parametersTink1);
    assertThat(parametersTink0.hashCode()).isEqualTo(parametersTink1.hashCode());
  }

  @Test
  public void testEqualsAndEqualHashCode_crunchy() throws Exception {
    Ed25519Parameters parametersCrunchy0 = Ed25519Parameters.create(CRUNCHY);
    Ed25519Parameters parametersCrunchy1 = Ed25519Parameters.create(CRUNCHY);
    assertThat(parametersCrunchy0).isEqualTo(parametersCrunchy1);
    assertThat(parametersCrunchy0.hashCode()).isEqualTo(parametersCrunchy1.hashCode());
  }

  @Test
  public void testEqualsAndEqualHashCode_legacy() throws Exception {
    Ed25519Parameters parametersLegacy0 = Ed25519Parameters.create(LEGACY);
    Ed25519Parameters parametersLegacy1 = Ed25519Parameters.create(LEGACY);
    assertThat(parametersLegacy0).isEqualTo(parametersLegacy1);
    assertThat(parametersLegacy0.hashCode()).isEqualTo(parametersLegacy1.hashCode());
  }

  @Test
  public void testNotEqualAndNotEqualHashCode_noPrefix() throws Exception {
    Ed25519Parameters parametersNoPrefix = Ed25519Parameters.create();

    Ed25519Parameters parametersTink = Ed25519Parameters.create(TINK);
    Ed25519Parameters parametersCrunchy = Ed25519Parameters.create(CRUNCHY);
    Ed25519Parameters parametersLegacy = Ed25519Parameters.create(LEGACY);

    assertThat(parametersNoPrefix).isNotEqualTo(parametersTink);
    assertThat(parametersNoPrefix.hashCode()).isNotEqualTo(parametersTink.hashCode());

    assertThat(parametersNoPrefix).isNotEqualTo(parametersCrunchy);
    assertThat(parametersNoPrefix.hashCode()).isNotEqualTo(parametersCrunchy.hashCode());

    assertThat(parametersNoPrefix).isNotEqualTo(parametersLegacy);
    assertThat(parametersNoPrefix.hashCode()).isNotEqualTo(parametersLegacy.hashCode());
  }

  @Test
  public void testNotEqualAndNotEqualHashCode_tink() throws Exception {
    Ed25519Parameters parametersTink = Ed25519Parameters.create(TINK);

    Ed25519Parameters parametersCrunchy = Ed25519Parameters.create(CRUNCHY);
    Ed25519Parameters parametersLegacy = Ed25519Parameters.create(LEGACY);

    assertThat(parametersTink).isNotEqualTo(parametersCrunchy);
    assertThat(parametersTink.hashCode()).isNotEqualTo(parametersCrunchy.hashCode());

    assertThat(parametersTink).isNotEqualTo(parametersLegacy);
    assertThat(parametersTink.hashCode()).isNotEqualTo(parametersLegacy.hashCode());
  }

  @Test
  public void testNotEqualAndNotEqualHashCode_crunchy_legacy() throws Exception {
    Ed25519Parameters parametersCrunchy = Ed25519Parameters.create(CRUNCHY);
    Ed25519Parameters parametersLegacy = Ed25519Parameters.create(LEGACY);

    assertThat(parametersCrunchy).isNotEqualTo(parametersLegacy);
    assertThat(parametersCrunchy.hashCode()).isNotEqualTo(parametersLegacy.hashCode());
  }
}

File: cc/signature/ed25519_parameters_test.cc
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/signature/ed25519_parameters.h"

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "tink/util/statusor.h"
#include "tink/util/test_matchers.h"

namespace crypto {
namespace tink {
namespace {

using ::crypto::tink::test::IsOk;
using ::crypto::tink::test::StatusIs;
using ::testing::Eq;
using ::testing::IsTrue;
using ::testing::TestWithParam;
using ::testing::Values;

struct TestCase {
  Ed25519Parameters::Variant variant;
  bool has_id_requirement;
};

using Ed25519ParametersTest = TestWithParam<TestCase>;

INSTANTIATE_TEST_SUITE_P(Ed25519ParametersTestSuite, Ed25519ParametersTest,
                         Values(TestCase{Ed25519Parameters::Variant::kTink,
                                         /*has_id_requirement=*/true},
                                TestCase{Ed25519Parameters::Variant::kCrunchy,
                                         /*has_id_requirement=*/true},
                                TestCase{Ed25519Parameters::Variant::kLegacy,
                                         /*has_id_requirement=*/true},
                                TestCase{Ed25519Parameters::Variant::kNoPrefix,
                                         /*has_id_requirement=*/false}));

TEST_P(Ed25519ParametersTest, Create) {
  TestCase test_case = GetParam();

  util::StatusOr<Ed25519Parameters> parameters =
      Ed25519Parameters::Create(test_case.variant);
  ASSERT_THAT(parameters, IsOk());

  EXPECT_THAT(parameters->GetVariant(), Eq(test_case.variant));
  EXPECT_THAT(parameters->HasIdRequirement(), Eq(test_case.has_id_requirement));
}

TEST(Ed25519ParametersTest, CreateWithInvalidVariantFails) {
  EXPECT_THAT(Ed25519Parameters::Create(
                  Ed25519Parameters::Variant::
                      kDoNotUseInsteadUseDefaultWhenWritingSwitchStatements)
                  .status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(Ed25519ParametersTest, CopyConstructor) {
  util::StatusOr<Ed25519Parameters> parameters =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kTink);
  ASSERT_THAT(parameters, IsOk());

  Ed25519Parameters copy(*parameters);

  EXPECT_THAT(copy.GetVariant(), Eq(Ed25519Parameters::Variant::kTink));
  EXPECT_THAT(copy.HasIdRequirement(), IsTrue());
}

TEST(Ed25519ParametersTest, CopyAssignment) {
  util::StatusOr<Ed25519Parameters> parameters =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kTink);
  ASSERT_THAT(parameters, IsOk());

  Ed25519Parameters copy = *parameters;

  EXPECT_THAT(copy.GetVariant(), Eq(Ed25519Parameters::Variant::kTink));
  EXPECT_THAT(copy.HasIdRequirement(), IsTrue());
}

TEST_P(Ed25519ParametersTest, ParametersEquals) {
  TestCase test_case = GetParam();

  util::StatusOr<Ed25519Parameters> parameters =
      Ed25519Parameters::Create(test_case.variant);
  ASSERT_THAT(parameters, IsOk());

  util::StatusOr<Ed25519Parameters> other_parameters =
      Ed25519Parameters::Create(test_case.variant);
  ASSERT_THAT(other_parameters, IsOk());

  EXPECT_TRUE(*parameters == *other_parameters);
  EXPECT_TRUE(*other_parameters == *parameters);
  EXPECT_FALSE(*parameters != *other_parameters);
  EXPECT_FALSE(*other_parameters != *parameters);
}

TEST(Ed25519ParametersTest, VariantNotEqual) {
  util::StatusOr<Ed25519Parameters> parameters =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kTink);
  ASSERT_THAT(parameters, IsOk());

  util::StatusOr<Ed25519Parameters> other_parameters =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kNoPrefix);
  ASSERT_THAT(other_parameters, IsOk());

  EXPECT_TRUE(*parameters != *other_parameters);
  EXPECT_FALSE(*parameters == *other_parameters);
}

}  // namespace
}  // namespace tink
}  // namespace crypto
